### ISSUE TITLE
lane arrow tool detection of bidirectional lane connections has a bug.

### ISSUE THREAD

**Issue by kianzarrin** [CitiesSkylinesMods/TMPE#738]
Lane arrow tool does not reset lane arrows for lanes that have lane connections.
but the code to detect lane connections has  a bug if the lane is bidirectional:
```C#
        /// <summary>
        /// determines whether or not the input lane is heading toward a start node.
        /// </summary>
        /// <returns>true if heading toward and start node.</returns>
        private bool IsHeadingTowardsStartNode(uint sourceLaneId) {
            NetLane[] laneBuffer = NetManager.instance.m_lanes.m_buffer;
            ushort segmentId = laneBuffer[sourceLaneId].m_segment;
            NetSegment segment = GetSeg(segmentId);
            uint laneId = segment.m_lanes;
            bool inverted = (segment.m_flags & NetSegment.Flags.Invert) != 0;

            foreach (var laneInfo in segment.Info.m_lanes) {
                if (laneId == sourceLaneId) {
///////////////////////////////
// what if laneInfo.m_direction is both
///////////////////////////////
                    return (laneInfo.m_direction == NetInfo.Direction.Forward) ^ !inverted;
                }
                laneId = laneBuffer[laneId].m_nextLane;
            }
            throw new Exception($"Unreachable code. sourceLaneId:{sourceLaneId}, segmentId:{segmentId} ");
        }
```
This means if the there is a segment with junctions at both ends, and one side of it has lane connections, it can confuse lane arrow tool.

**Comment by aubergine10** [comment]
I imagine this is particularly problematic when it comes to things like single train tracks that allow bidirectional traffic with the[ Single Train Track AI](https://steamcommunity.com/sharedfiles/filedetails/?id=949504539)?

**Comment by kianzarrin** [comment]
Lane arrow tool does not support tracks

**Comment by aubergine10** [comment]
> Network extension 2 has some roads with turning lane. the turning lane is bidirectional. For a small segment I have seen cars moving on it in both directions going through each other! I don't know if mass transit DLC would fix this.

No, Mass Transit doesn't resolve those issues. Maybe we need a "Single Road Track AI" 😄 

From vague memory there are a few other workshop road assets that use 'both' for their "turning lanes", although many use a different trick: Two directional lanes (one forward, one backward) on top of each other lol.

> Lane arrow tool does not support tracks

True, although once #734 is done we could potentially add lane arrow suport as a "direction-based alternative" to lane connectors?

**Comment by kianzarrin** [comment]
Silly me! I should have tested things before speculating.
I tested my code is working fine. NExt2 does not use bidirectional lanes but rather two single directional lanes on the top of each other.

If any road does use bidirectional road lanes (i don't even know if that is possible) then TMPE will be unable to store the flags for both ends (because there is only memory allocated for one lane).

**Comment by kianzarrin** [comment]
the NExt2 turning lane only has left turn prop but cars do go straight:
![Screenshot (892)](https://user-images.githubusercontent.com/26344691/79703524-ff780000-82b4-11ea-9814-35dee812b141.png)
![Screenshot (890)](https://user-images.githubusercontent.com/26344691/79703516-f5ee9800-82b4-11ea-95a6-16d35c12c72e.png)
but that is an issue for NExt2

**Comment by kianzarrin** [comment]
I found a new issue though. I cannot reset lane arrows after I remove the lane connection.

**Comment by aubergine10** [comment]
Is that only for NExt2 roads or all roads?
